url2bc31e3489_ramonli_RabbitMQResearch_tgz-pJ8-com_mpos_lottery_te_thirdpartyservice_amqp_protobuf_CancellationMessagePulisherJ8/src/cluster/NewSend.java

```diff
@@ -0,0 +1,79 @@
+package cluster;
+
+import com.rabbitmq.client.Address;
+import com.rabbitmq.client.Channel;
+import com.rabbitmq.client.Connection;
+import com.rabbitmq.client.ConnectionFactory;
+
+/**
+ * > ConnectionFactory factory = new ConnectionFactory();<br/>
+ * > Address[] addresses = {new Address("localhost", 12345), new <br/>
+ * > Address("localhost", 12346)}; > factory.newConnection(addresses); <br/>
+ * > <br/>
+ * > If I'll do this, how can I know what node of cluster I'm connected?
+ * <p>
+ * It shouldn't matter which node you are connected to (unless you care about
+ * the locations of queues). Clients see an identical view of the world
+ * regardless of which broker in the same cluster they connect to.
+ * <p>
+ * A channel can query the underlying connection. You can find the address of
+ * the clustered node that way if necessary.
+ * <p>
+ * > If node 1 down, but node 2 still up, how can I reconnect in node 1?
+ * <p>
+ * The factory connection does not offer robust reconnection logic. If you want
+ * clients to reconnect then take a look at the message patterns library:
+ * <p>
+ * http://hg.rabbitmq.com/rabbitmq-java-messagepatterns
+ * <p>
+ * This is a higher-level library that offers reconnection logic, which is
+ * sufficiently flexible for use with clusters.
+ * <p>
+ * Refer to https://groups.google.com/forum/#!topic/rabbitmq-discuss/Qvbff7jnAC4
+ * 
+ * @author Ramon
+ * 
+ */
+public class NewSend {
+    private final static String QUEUE_NAME = "hello";
+
+    public static void main(String[] argv) throws java.io.IOException {
+        // create connection
+        ConnectionFactory factory = new ConnectionFactory();
+        // setup a cluster contains 2 nodes[new Address("localhost", 5672), new Address("localhost", 5673)]
+        // refer to "A cluster on a single machine"#http://www.rabbitmq.com/clustering.html
+        Connection connection = factory.newConnection(new Address[] { new Address("localhost", 5672)});        
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *connection*. The resource is referenced at line 46. The resource is closed at line 59. There are other execution paths that do not contain closure statements, for example, when Connection.createChannel throws an exception. Ensure *connection* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).